File: 0101-0200/0152.maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int pos = 1, neg = 1;
        int ans = nums[0];
        for (int &x : nums) {
            if (x == 0) {
                pos = 0; neg = 0;
            } else if (x > 0) {
                pos = max(x, pos * x);
                neg = neg * x;
            } else {
                int old_pos = pos;
                pos = neg * x;
                neg = min(x, x * old_pos);
            }
            ans = max(ans, pos);
        }
        return ans;
    }
};

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = nums[0];
        const int size = nums.size();
        vector<int> min_arr(size), max_arr(size);
        min_arr[0] = max_arr[0] = nums[0];
        for (int i = 1; i < size; ++i) {
            int x = nums[i];
            int a = min_arr[i-1] * x;
            int b = max_arr[i-1] * x;
            min_arr[i] = min({x, a, b});
            max_arr[i] = max({x, a, b});
            ans = max(max_arr[i], ans);
        }
        return ans;
    }
};

File: 0101-0200/0152-maximum-product-subarray.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = nums[0]
        pos = 0
        neg = 0
        for x in nums:
            if x == 0:
                ans = max(ans, 0)
                pos = 0
                neg = 0
            elif x > 0:
                pos = max(x, pos * x)
                neg = neg * x
            else:
                # 注意min/max
                pos, neg = neg * x, min(pos * x, x)
            if pos != 0:
                ans = max(ans, pos)
            if neg != 0:
                ans = max(ans, neg)
        return ans

print(Solution().maxProduct([2,3,-2,4]))
print(Solution().maxProduct([-2, 0, -1]))
print(Solution().maxProduct([-3, -1, -1]))
